### PR TITLE
Add wsd support

### PR DESCRIPTION
**Pull request by tomtang502**
- Add WSD (warmup-stable-decay) LR schedule: scheduler="WSD" holds max_lr, then decays to min_lr over the last decay_steps, shape picked by decay_type (linear/cosine).
- Validate decay_steps (0 < decay_steps < max_steps − warmup_steps) so a bad value fails fast at startup instead of crashing near the end or silently pinning the LR.
- Fix warmup_steps=0 — it used to hold the whole run at min_lr; now it starts at max_lr (also fixes Constant/Cosine, and makes warmup-free decay-from-a-checkpoint work).
- Resuming a checkpoint under a different scheduler shape now raises a clear error instead of an IndexError / silently-wrong LR. 
- CosineAnnealing/Constant paths unchanged; verified across pp/dp/cp/ep meshes and both decay shapes.

### CODE REVIEW

File: pithtrain/tasks/pretrain_lm.py [hunk 1]
```diff
@@ -224,6 +224,18 @@ def load_state_dict(self, state_dict):
...
+            # shaped scheduler silently corrupts the LR (or raises IndexError), so refuse a
+            # structural mismatch with a clear message instead.
+            saved_n = len(sched_state.get("_schedulers", []))
+            cur_n = len(getattr(self.scheduler, "_schedulers", []))
+            if saved_n != cur_n:
+                raise ValueError(
+                    f"Checkpoint scheduler has {saved_n} phase(s) but the current config "
+                    f"builds {cur_n}; resuming across a scheduler/warmup/decay change is not "
+                    f"supported. Resume with the same scheduler config, or train from scratch."
+                )
```
[the remark this answers is not in the document]
Comment (tomtang502) [reply pithtrain/tasks/pretrain_lm.py]: agree, added a recursive comparison

File: pithtrain/modules/training.py [hunk 2]
```diff
@@ -412,15 +419,41 @@ def setup_optimizer(cfg: TrainingCfg, ctx: TrainingCtx) -> None:
...
+                    f"WSD needs 0 < decay_steps < max_steps - warmup_steps; got "
+                    f"decay_steps={cfg.decay_steps}, warmup_steps={warmup_steps}, max_steps={max_steps}"
+                )
+            stable_steps = post_steps - cfg.decay_steps
+            phases.append(LinearLR(ctx.optimizer, 1.0, 1.0, stable_steps))
+            phases.append(
+                LinearLR(ctx.optimizer, 1.0, min_lr / max_lr, cfg.decay_steps)
+                if cfg.decay_type == "linear"
+                else CosineAnnealingLR(ctx.optimizer, cfg.decay_steps, min_lr)
+            )
```
[the remark this answers is not in the document]
Comment (tomtang502) [reply pithtrain/modules/training.py]: agree, refactor with prev issue

File: pithtrain/modules/training.py [hunk 3]
```diff
@@ -412,15 +419,41 @@ def setup_optimizer(cfg: TrainingCfg, ctx: TrainingCtx) -> None:
 def setup_scheduler(cfg: TrainingCfg, ctx: TrainingCtx) -> None:
     min_lr, max_lr = cfg.min_lr, cfg.max_lr
     warmup_steps, max_steps = cfg.warmup_steps, cfg.max_steps
-    warmup = LinearLR(ctx.optimizer, min_lr / max_lr, 1.0, warmup_steps)
+    post_steps = max_steps - warmup_steps
```
[the remark this answers is not in the document]
Comment (tomtang502) [reply pithtrain/modules/training.py:423]: resolved

File: pithtrain/modules/training.py [hunk 4]
```diff
@@ -412,15 +419,48 @@ def setup_optimizer(cfg: TrainingCfg, ctx: TrainingCtx) -> None:
...
+                    phases.append(CosineAnnealingLR(ctx.optimizer, cfg.decay_steps, min_lr))
+                case _:
+                    raise ValueError(f"Unknown decay_type: {cfg.decay_type!r}")
+            milestones.append(warmup_steps + stable_steps)
         case _:
             raise ValueError(f"Unknown scheduler: {cfg.scheduler!r}")
-    ctx.scheduler = SequentialLR(ctx.optimizer, [warmup, stable], [warmup_steps])
+    ctx.scheduler = (
+        SequentialLR(ctx.optimizer, phases, milestones) if len(phases) > 1 else phases[0]
+    )
```
Comment (haok1402) [inline pithtrain/modules/training.py]: @tomtang502 We don't need an extra if-else here ; `SequentialLR` would chain a singleton too ; 

```python3
  inner = LinearLR(opt, start_factor=0.5, total_iters=5)
  sched = SequentialLR(opt, schedulers=[inner], milestones=[])
```

File: pithtrain/tasks/pretrain_lm.py [hunk 5]
```diff
@@ -163,6 +163,23 @@ def clip_grad_norm_(model: nn.Module, max_norm: float, norm_type: float = 2.0) -
     return total_norm
 
 
+def _scheduler_structure_mismatch(state: dict, sched: LRScheduler) -> bool:
```
Comment (haok1402) [inline pithtrain/tasks/pretrain_lm.py:166]: @tomtang502 Should this check the type of that scheduler too?

File: pithtrain/tasks/pretrain_lm.py [hunk 6]
```diff
@@ -224,6 +241,16 @@ def load_state_dict(self, state_dict):
             options = StateDictOptions(strict=False)
             set_model_state_dict(self.model, model_state, options=options)
         if sched_state:
+            # A scheduler's structure (phase count AND per-phase type) depends on the config
+            # (warmup / scheduler / WSD decay shape). torch load_state_dict just updates
+            # __dict__, so loading a mismatched structure silently corrupts the LR or crashes;
+            # refuse any structural/type mismatch with a clear message instead.
```
Comment (haok1402) [inline pithtrain/tasks/pretrain_lm.py]: @tomtang502 nit: excessive comments here ; The code below from L248 - L253 is pretty self-explanatory.

File: pithtrain/modules/training.py [hunk 7]
```diff
@@ -65,9 +65,16 @@ class TrainingCfg(SlottedDefault):
...
-    scheduler: Literal["CosineAnnealing", "Constant"]
+    scheduler: Literal["CosineAnnealing", "Constant", "WSD"]
     """The learning rate scheduler to use after linear warmup."""
 
+    decay_type: Literal["linear", "cosine"] = "cosine"
+    """Shape of the WSD decay tail. Only used when ``scheduler == "WSD"``."""
+
+    decay_steps: int = 0
+    """Number of steps in the WSD decay tail; the stable phase is the remainder
+    (``max_steps - warmup_steps - decay_steps``). Only used when ``scheduler == "WSD"``."""
```
Comment (haok1402) [inline pithtrain/modules/training.py]: @tomtang502 If these fields only apply to WSD, should their field names contain WSD too? Otherwise, one may very much claim that the overall decay can be cosine by setting up `decay_type` ; yet that's not what this field is intended.

File: pithtrain/modules/training.py [hunk 8]
```diff
@@ -412,15 +419,48 @@ def setup_optimizer(cfg: TrainingCfg, ctx: TrainingCtx) -> None:
...
+    if warmup_steps > 0:
+        phases.append(LinearLR(ctx.optimizer, min_lr / max_lr, 1.0, warmup_steps))
+        milestones.append(warmup_steps)
     match cfg.scheduler:
         case "CosineAnnealing":
-            stable = CosineAnnealingLR(ctx.optimizer, max_steps - warmup_steps, min_lr)
+            phases.append(CosineAnnealingLR(ctx.optimizer, post_steps, min_lr))
         case "Constant":
-            stable = LinearLR(ctx.optimizer, 1.0, 1.0, max_steps - warmup_steps)
+            phases.append(LinearLR(ctx.optimizer, 1.0, 1.0, post_steps))
```
Comment (haok1402) [inline pithtrain/modules/training.py:439]: @tomtang502 I wonder if we should set a default of `0` to the warmup steps? By doing this, people using `ConstantLR` scheduler don't need to further specify `warmup_steps=0`.

File: pithtrain/modules/training.py (hunk 4 above)
Comment (tomtang502) [reply pithtrain/modules/training.py]: refatored

File: pithtrain/tasks/pretrain_lm.py (hunk 5 above)
Comment (tomtang502) [reply pithtrain/tasks/pretrain_lm.py:166]: checked

File: pithtrain/modules/training.py (hunk 7 above)
Comment (tomtang502) [reply pithtrain/modules/training.py]: add wsd preffix

File: pithtrain/tasks/pretrain_lm.py (hunk 6 above)
Comment (tomtang502) [reply pithtrain/tasks/pretrain_lm.py]: removed

File: pithtrain/modules/training.py (hunk 8 above)
Comment (tomtang502) [reply pithtrain/modules/training.py:439]: agree, added

**Review by haok1402 — APPROVED** [review]
